File: src/memory_agent/extractor.py
"""Extractor: post-conversation memory extraction with user review."""
import json
import uuid
from dataclasses import dataclass, field
from datetime import datetime, timezone
import httpx
from memory_agent.config import Config
from memory_agent.storage import MemoryStore
from memory_agent.prompts import EXTRACTOR_SYSTEM_PROMPT, EXTRACTOR_USER_TEMPLATE


@dataclass
class ExtractionResult:
    summary: str = ""
    key_points: list[str] = field(default_factory=list)
    tags: list[str] = field(default_factory=list)
    entities: list[dict] = field(default_factory=list)
    decisions: list[str] = field(default_factory=list)

    @classmethod
    def from_dict(cls, data: dict) -> "ExtractionResult":
        return cls(
            summary=data.get("summary", ""), key_points=data.get("key_points", []),
            tags=data.get("tags", []), entities=data.get("entities", []),
            decisions=data.get("decisions", []),
        )


def _call_extraction_llm(config: Config, transcript: str) -> ExtractionResult:
    response = httpx.post(
        f"{config.llm_api_base}/chat/completions",
        headers={"Authorization": f"Bearer {config.llm_api_key}", "Content-Type": "application/json"},
        json={
            "model": config.llm_model,
            "messages": [
                {"role": "system", "content": EXTRACTOR_SYSTEM_PROMPT},
                {"role": "user", "content": EXTRACTOR_USER_TEMPLATE.format(transcript=transcript)},
            ],
            "temperature": 0.3, "max_tokens": 1000,
        },
        timeout=60,
    )
    response.raise_for_status()
    data = response.json()
    content = data["choices"][0]["message"]["content"]
    return ExtractionResult.from_dict(json.loads(content))


def _display_preview(result: ExtractionResult) -> None:
    print("\n" + "=" * 50)
    print("📝 Memory Preview")
    print("=" * 50)
    print(f"\nSummary: {result.summary}")
    print(f"\nTags: {', '.join(result.tags) or '(none)'}")
    print(f"\nKey Points:")
    for kp in result.key_points:
        print(f"  • {kp}")
    if not result.key_points:
        print("  (none)")
    print(f"\nEntities:")
    for ent in result.entities:
        print(f"  • {ent['name']} ({ent['type']}): {ent.get('description', '')}")
    if not result.entities:
        print("  (none)")
    print(f"\nDecisions:")
    for dec in result.decisions:
        print(f"  • {dec}")
    if not result.decisions:
        print("  (none)")
    print()


def _get_user_choice() -> str:
    while True:
        choice = input("[S]ave  [E]dit  [D]iscard: ").strip().lower()
        if choice in ("s", "save", "y", "yes"): return "save"
        elif choice in ("d", "discard", "n", "no"): return "discard"
        elif choice in ("e", "edit"): return "edit"
        else: print("Please enter S, E, or D")


def _open_editor(result: ExtractionResult) -> ExtractionResult:
    import subprocess, tempfile, os
    editor = os.environ.get("EDITOR", "vim")
    data = {"summary": result.summary, "key_points": result.key_points, "tags": result.tags, "entities": result.entities, "decisions": result.decisions}
    content = json.dumps(data, indent=2, ensure_ascii=False)
    with tempfile.NamedTemporaryFile(mode="w", suffix=".json", delete=False) as f:
        f.write(content)
        tmp_path = f.name
    try:
        subprocess.run([editor, tmp_path], check=False)
        with open(tmp_path) as f:
            edited = json.loads(f.read())
        return ExtractionResult.from_dict(edited)
    finally:
        os.unlink(tmp_path)


def extract_and_store(transcript: str, config: Config, store: MemoryStore) -> bool:
    print("\nExtracting memory from conversation...")
    try:
        result = _call_extraction_llm(config, transcript)
    except Exception as e:
        print(f"Extraction failed: {e}")
        return False

    if config.extractor_auto_confirm:
        _store_result(result, transcript, config, store)
        print("Memory saved (auto-confirm).")
        return True

    while True:
        _display_preview(result)
        choice = _get_user_choice()
        if choice == "save":
            _store_result(result, transcript, config, store)
            print("Memory saved.")
            return True
        elif choice == "edit":
            result = _open_editor(result)
        elif choice == "discard":
            print("Memory discarded.")
            return False


def _store_result(result: ExtractionResult, transcript: str, config: Config, store: MemoryStore) -> None:
    now = datetime.now(timezone.utc)
    chroma_dir = config.memory_dir / "chroma"
    if not hasattr(store, "_chroma_collection") or store._chroma_collection is None:
        store.init_chroma(persist_dir=chroma_dir, embedding_api_base=config.embedding_api_base, embedding_api_key=config.embedding_api_key, embedding_model=config.embedding_model)
    embedding_text = result.summary
    if result.key_points:
        embedding_text += "\n" + "\n".join(result.key_points)
    memory_id = uuid.uuid4().hex[:12]
    chroma_doc_id = store.add_to_chroma(memory_id=memory_id, text=embedding_text, metadata={"tags": ",".join(result.tags), "conversation_at": now.isoformat()})
    conversation_json = json.dumps({"transcript": transcript}) if config.extractor_keep_full_transcript else None
    memory_id = store.insert_memory(
        summary=result.summary, conversation_at=now, conversation_json=conversation_json,
        chroma_doc_id=chroma_doc_id, key_points=result.key_points, tags=result.tags,
        entities=result.entities, decisions=result.decisions,
        memory_id=memory_id,
    )
    print(f"  Memory ID: {memory_id}")

File: src/extractor.rs
use crate::config::Config;
use crate::prompts::{EXTRACTOR_SYSTEM_PROMPT, EXTRACTOR_USER_TEMPLATE};
use crate::storage::MemoryStore;
use anyhow::Result;
use serde::{Deserialize, Serialize};
/// Extractor: post-conversation memory extraction with user review.
use std::io::{self, Write};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ExtractionResult {
    pub summary: String,
    #[serde(default)]
    pub key_points: Vec<String>,
    #[serde(default)]
    pub tags: Vec<String>,
    #[serde(default)]
    pub entities: Vec<serde_json::Value>,
    #[serde(default)]
    pub decisions: Vec<String>,
}

async fn call_extraction_llm(config: &Config, transcript: &str) -> Result<ExtractionResult> {
    let client = reqwest::Client::new();
    let url = format!("{}/chat/completions", config.llm_api_base);

    let resp = client
        .post(&url)
        .header("Authorization", format!("Bearer {}", config.llm_api_key))
        .header("Content-Type", "application/json")
        .json(&serde_json::json!({
            "model": config.llm_model,
            "messages": [
                {"role": "system", "content": EXTRACTOR_SYSTEM_PROMPT},
                {"role": "user", "content":
                    EXTRACTOR_USER_TEMPLATE.replace("{transcript}", transcript)},
            ],
            "temperature": 0.3,
            "max_tokens": 1000,
        }))
        .send()
        .await?
        .error_for_status()?;

    let data: serde_json::Value = resp.json().await?;
    let content = data["choices"][0]["message"]["content"]
        .as_str()
        .unwrap_or("{}");

    Ok(serde_json::from_str(content)?)
}

fn display_preview(result: &ExtractionResult) {
    eprintln!("\n==================================================");
    eprintln!("\u{1F4DD} Memory Preview");
    eprintln!("==================================================");
    eprintln!("\nSummary: {}", result.summary);
    eprintln!(
        "\nTags: {}",
        if result.tags.is_empty() {
            "(none)".to_string()
        } else {
            result.tags.join(", ")
        }
    );
    eprintln!("\nKey Points:");
    if result.key_points.is_empty() {
        eprintln!("  (none)");
    } else {
        for kp in &result.key_points {
            eprintln!("  \u{2022} {kp}");
        }
    }
    eprintln!("\nEntities:");
    if result.entities.is_empty() {
        eprintln!("  (none)");
    } else {
        for ent in &result.entities {
            eprintln!(
                "  \u{2022} {} ({}): {}",
                ent["name"].as_str().unwrap_or(""),
                ent["type"].as_str().unwrap_or(""),
                ent["description"].as_str().unwrap_or(""),
            );
        }
    }
    eprintln!("\nDecisions:");
    if result.decisions.is_empty() {
        eprintln!("  (none)");
    } else {
        for dec in &result.decisions {
            eprintln!("  \u{2022} {dec}");
        }
    }
    eprintln!();
}

fn get_user_choice() -> String {
    loop {
        eprint!("[S]ave  [E]dit  [D]iscard: ");
        let _ = io::stderr().flush();
        let mut input = String::new();
        if io::stdin().read_line(&mut input).is_err() {
            return "discard".to_string();
        }
        match input.trim().to_lowercase().as_str() {
            "s" | "save" | "y" | "yes" => return "save".to_string(),
            "d" | "discard" | "n" | "no" => return "discard".to_string(),
            "e" | "edit" => return "edit".to_string(),
            _ => eprintln!("Please enter S, E, or D"),
        }
    }
}

fn open_editor(result: &ExtractionResult) -> ExtractionResult {
    let editor = std::env::var("EDITOR").unwrap_or_else(|_| "vim".to_string());
    let json_str = serde_json::to_string_pretty(result).unwrap_or_default();
    let tmp_path =
        std::env::temp_dir().join(format!("memory_extract_{}.json", uuid::Uuid::new_v4()));

    if std::fs::write(&tmp_path, &json_str).is_err() {
        return result.clone();
    }

    // Launch editor
    let _ = std::process::Command::new(&editor).arg(&tmp_path).status();

    // Read back
    if let Ok(edited) = std::fs::read_to_string(&tmp_path) {
        if let Ok(r) = serde_json::from_str(&edited) {
            let _ = std::fs::remove_file(&tmp_path);
            return r;
        }
    }
    let _ = std::fs::remove_file(&tmp_path);
    result.clone()
}

pub async fn extract_and_store(
    transcript: &str,
    config: &Config,
    store: &MemoryStore,
    auto_confirm: Option<bool>,
) -> Result<bool> {
    eprintln!("\nExtracting memory from conversation...");

    let result = match call_extraction_llm(config, transcript).await {
        Ok(r) => r,
        Err(e) => {
            eprintln!("Extraction failed: {e}");
            return Ok(false);
        }
    };

    let effective_auto = auto_confirm.unwrap_or(config.extractor_auto_confirm);

    if effective_auto {
        store_result(&result, transcript, config, store).await?;
        eprintln!("Memory saved (auto-confirm).");
        return Ok(true);
    }

    let mut current = result;
    loop {
        display_preview(&current);
        match get_user_choice().as_str() {
            "save" => {
                store_result(&current, transcript, config, store).await?;
                eprintln!("Memory saved.");
                return Ok(true);
            }
            "edit" => {
                current = open_editor(&current);
            }
            "discard" => {
                eprintln!("Memory discarded.");
                return Ok(false);
            }
            _ => unreachable!(),
        }
    }
}

async fn store_result(
    result: &ExtractionResult,
    transcript: &str,
    config: &Config,
    store: &MemoryStore,
) -> Result<()> {
    let now = chrono::Utc::now();

    // Build embedding text from summary + key points
    let mut embedding_text = result.summary.clone();
    if !result.key_points.is_empty() {
        embedding_text.push('\n');
        embedding_text.push_str(&result.key_points.join("\n"));
    }

    let memory_id: String = uuid::Uuid::new_v4().to_string().chars().take(12).collect();

    // Build conversation JSON
    let conversation_json = if config.extractor_keep_full_transcript {
        Some(serde_json::json!({"transcript": transcript}).to_string())
    } else {
        None
    };

    // Insert metadata first, then the vector (which sets memories.vec_rowid).
    store.insert_memory(
        &result.summary,
        &now,
        conversation_json.as_deref(),
        &result.key_points,
        &result.tags,
        &result.entities,
        &result.decisions,
        Some(&memory_id),
    )?;
    store.add_memory_vector(&memory_id, &embedding_text).await?;

    eprintln!("  Memory ID: {memory_id}");
    Ok(())
}
